x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -1282,6 +1288,11 @@ public void onIndexModule(IndexModule module) {
         );
     }
 
+    @Override
+    public void onNodeStarted() {
+        this.nodeStartedSignal.countDown();
```

Comment:
Is there a reason we signal rather than just moving the call to `InitialNodeSecurityAutoConfiguration.maybeGenerateEnrollmentTokensAndElasticCredentialsOnNodeStartup` into this method?

I presume it's because we'd need to push all the arguments somewhere, but with this approach we've got a generic thread sitting waiting on a countdown when it could just be deferred until the right part of the lifecycle.

I understand that we're trying to do the smallest thing here, so maybe it's OK for now, but I think it would be better to make `InitialNodeSecurityAutoConfiguration` a real object, and store it in a field, and then call `execute()` on it from `onNodeStarted`.

Reply:
> Is there a reason we signal rather than just moving the call to InitialNodeSecurityAutoConfiguration.maybeGenerateEnrollmentTokensAndElasticCredentialsOnNodeStartup into this method?

`securityIndexManager.onStateRecovered` must be invoked before the cluster state recovers, whcih means somewhere in the `Node` constructor.

Reply:
To be clearer, there are technically two conditions that must be met to trigger this last part of auto-configuration:
 1) the cluster state is just recovered and does not contain the `.security` index (which might later be created by other processes)
 2) the node is started
 
In practice, in my tests, these happen almost simultaneously.

In order to detect 1), a Security Index manager listener has to be registered before cluster state recovery starts, otherwise it might miss the notification for the state recovery.

Reply:
>I understand that we're trying to do the smallest thing here, so maybe it's OK for now, but I think it would be better to make InitialNodeSecurityAutoConfiguration a real object, and store it in a field, and then call execute() on it from onNodeStarted.

I have replaced the `CountdownLatch` with a `ListenableFuture`, 4902844 .
This way no thread is blocked, instead a listener is registered, which can be called by anyone on any thread because itself only schedules an action.